content/blogposts/2023/slow-to-simd.md

```diff
@@ -0,0 +1,411 @@
+            panic("out of bounds")
+        }
+		s2 := a[i+2] * b[i+2]
+        if i+3 >= cap(a) || >= cap(b) {
+            panic("out of bounds")
+        }
+		s3 := a[i+3] * b[i+3]
+		sum += s0 + s1 + s2 + s3
+	}
+	return sum
+}
+```
+
+In a hot loop like this, even with modern branch prediction, the additional branches per iteration can add up to a pretty significant performance penalty. This is especially true in our case because the inserted jumps limit how much we can take advantage of pipelining.
+
+If we can convince the compiler that these reads can never be out of bounds, it won't insert these runtime checks. This technique is known as "bounds-checking elimination", and the same pattern can be applied to many different memory-safe compiled languages.
+
+In theory, we should be able to do all checks once, outside the loop, and the compiler would be able to determine that all the slice indexing is safe. However, I couldn't find the right combination of checks to convince the compiler that what I'm doing is safe. I landed on a combination of asserting the lengths are equal and moving all the bounds checking to the top of the loop. This was enough to hit nearly the speed of the bounds-check-free version.
+
+```go
+func DotBCE(a, b []float32) float32 {
+	if len(a) != len(b) {
+		panic("slices must have equal lengths")
+	}
+
+    if len(a)%4 != 0 {
+		panic("slice length must be multiple of 4")
+	}
+
+	sum := float32(0)
+	for i := 0; i < len(a); i += 4 {
+		aTmp := a[i : i+4 : i+4]
+		bTmp := b[i : i+4 : i+4]
+		s0 := aTmp[0] * bTmp[0]
+		s1 := aTmp[1] * bTmp[1]
+		s2 := aTmp[2] * bTmp[2]
+		s3 := aTmp[3] * bTmp[3]
+		sum += s0 + s1 + s2 + s3
+	}
+	return sum
+}
+```
+
+The minimizing of bounds checking nets a 9% improvement. Consistently non-zero, but nothing to write home about. 
+
+<Chart
+    rows={[
+        {name: "DotNaive", value: 941900},
+        {name: "DotUnroll4", value: 1286000},
+        {name: "DotBCE", value: 1407000},
+    ]}
+/>
+
+This technique translates well to many memory-safe compiled languages like [Rust](https://nnethercote.github.io/perf-book/bounds-checks.html).
+
+Exercise for the reader: why is it significant that we slice like `a[i:i+4:i+4]` rather than just `a[i:i+4]`?
+
+## Quantization
+
+We've improved execution speed of our code pretty dramatically at this point, but there is another dimension of performance that is relevant here: memory usage.
+
+(If you skipped the background section, this part might not make much sense.)
+
+In our situation, we're searching over vectors with 1536 dimensions. With 4-byte elements, this comes out to 6KiB per vector, and we generate roughly a million vectors per GiB of code. That adds up.
+
+When searching the vectors, they need to be held in RAM, which puts some serious memory pressure on our deployments. Moving the vectors out of memory would mean loading them from disk at search time, which is a no-go when performance is so critical.
```

Comment:
I think it's worth motivating why disks are so much slower in memory in this case. Maybe talk about https://colin-scott.github.io/personal_website/research/interactive_latency.html (or an example based off of this). 

( Should mention that we're doing a lot of random reads...)

Reply:
Oh, that's a great link! I'll incorporate that